### PR TITLE
Enhance e2e to pick non-Gateway nodes via labels

### LINKED ISSUE #1244: Enhance e2e to pick non-Gateway nodes via labels
Currently, `subctl verify ...` picks random non-Gateway nodes to verify various use-cases. However, in large deployments if there is any connectivity issue on a specific node, the tool will not be able to identify the issue unless the client/server pod is scheduled on the node.

To troubleshoot connectivity issues, we can enhance Shipyard framework to look for nodes labelled with a specific value (say `submariner.io/e2e-non-gw-node=true`) and use those nodes as non-Gateway nodes while validating datapath use-cases. If the label is not added to any of the nodes, then the existing mechanism can be used.

### PR DESCRIPTION
**Pull request by sridhargaddam**
Currently, subctl verify ... picks random non-Gateway nodes to verify various use-cases. However, in large deployments if there is any connectivity issue on a specific node, the tool will not be able to identify the issue unless the client/server pod is scheduled on the node.

This PR enhances the code to select non-Gateway nodes specifically labeled with `test.submariner.io/non-gateway-node=true` while validating datapath use-cases. If none of the nodes have the label, the test framework falls back to the existing approach of randomly selecting a non-Gateway node.

Fixes: https://github.com/submariner-io/shipyard/issues/1244

### CODE REVIEW

File: release-notes/20230508-e2e-non-gw-nodes.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<!-- markdownlint-disable MD041 -->
+The e2e test framework has been enhanced to select nodes labeled with `submariner.io/e2e-non-gw-node=true` as non-Gateway nodes
```
Comment (tpantelis) [inline release-notes/20230508-e2e-non-gw-nodes.md]: `subctl` is really the user-facing aspect.

```suggestion
`subctl verify` has been enhanced to select nodes labeled with `submariner.io/e2e-non-gw-node=true` as non-Gateway nodes
```

File: test/e2e/framework/framework.go [hunk 2]
```diff
@@ -65,6 +65,7 @@ const (
 	RouteAgent         = "submariner-routeagent"
 	GatewayLabel       = "submariner.io/gateway"
 	ActiveGatewayLabel = "gateway.submariner.io/status=active"
+	E2eNonGWNodeLabel  = "submariner.io/e2e-non-gw-node=true"
```
Comment (tpantelis) [inline test/e2e/framework/framework.go]:
```suggestion
	TestNonGWNodeLabel  = "test.submariner.io/non-gateway-node=true"
```

File: test/e2e/framework/network_pods.go [hunk 3]
```diff
@@ -554,22 +554,38 @@ func (np *NetworkPod) nodeAffinity(scheduling NetworkPodScheduling) *v1.Affinity
...
+
+		switch {
+		case len(nonGWNodes) > 0:
+			nodeSelTerms = addNodeSelectorTerm(nodeSelTerms, "kubernetes.io/hostname", v1.NodeSelectorOpIn, nonGWNodes)
+		case len(smE2eNonGWLabelledNodeList.Items) > 0 && len(nonGWNodes) == 0:
+			Failf("%q label is only present on the active GW node and not on any other nodes", E2eNonGWNodeLabel)
+		default:
+			nodeSelTerms = addNodeSelectorTerm(nodeSelTerms, GatewayLabel, v1.NodeSelectorOpDoesNotExist, nil)
+			nodeSelTerms = addNodeSelectorTerm(nodeSelTerms, GatewayLabel, v1.NodeSelectorOpNotIn, []string{"true"})
+		}
```
Comment (tpantelis) [inline test/e2e/framework/network_pods.go]: I think this simplifies it a bit, since we know `len(nonGWNodes) == 0`:

```suggestion
		case len(smE2eNonGWLabelledNodeList.Items) == 0:
		        nodeSelTerms = addNodeSelectorTerm(nodeSelTerms, GatewayLabel, v1.NodeSelectorOpDoesNotExist, nil)
			nodeSelTerms = addNodeSelectorTerm(nodeSelTerms, GatewayLabel, v1.NodeSelectorOpNotIn, []string{"true"})
		default:
		        Failf("%q label is only present on the active GW node and not on any other nodes", E2eNonGWNodeLabel)
		}
```

File: test/e2e/framework/network_pods.go [hunk 4]
```diff
@@ -584,6 +600,19 @@ func (np *NetworkPod) nodeAffinity(scheduling NetworkPodScheduling) *v1.Affinity
 	}
 }
 
+func (np *NetworkPod) activeGatewayHostname() string {
+	smGWPodList, err := KubeClients[np.Config.Cluster].CoreV1().Pods(TestContext.SubmarinerNamespace).List(context.TODO(),
+		metav1.ListOptions{LabelSelector: ActiveGatewayLabel})
+
+	Expect(err).NotTo(HaveOccurred())
+	Expect(smGWPodList.Items).To(HaveLen(1))
```
Comment (tpantelis) [inline test/e2e/framework/network_pods.go]: The original code used `AwaitUntil` - why was that removed?

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Comment by sridhargaddam** [comment]
@tpantelis as the changes were not many, instead of using another commit, I appended the changes to the same commit.

**Review by yboaron — APPROVED** [review]

**Review by dfarrell07 — APPROVED** [review]

**Comment by sridhargaddam** [comment]
The `Linting/Shell` failures are not related to this PR. @dfarrell07 is this some known issue and someone looking into it?

**Comment by skitt** [comment]
> The `Linting/Shell` failures are not related to this PR. @dfarrell07 is this some known issue and someone looking into it?

These should be fixed by #1247.

**Review by skitt — APPROVED** [review]

**Comment by sridhargaddam** [comment]
@tpantelis addressed your comments, can you pls take a relook?

**Review by tpantelis — APPROVED** [review]
